Clang: Substantially rework CAS-related command-line options

dexonsmith opened the pull request:
This substantially reworks the command-line options in Clang related to
the CAS, primarly in order to strip down `CASOptions` to only contain
CAS configuration, and not indications of how to use it. It also
fine-tuned how to configure the CAS, which caused a long-tail of things
layering problems to be pulled in. I cut off a few of the threads with
FIXMEs for now.

The new CASOptions only has a single field, BuiltinPath, which is stored
in a sub-structure called CASConfiguration.

- There is no longer an `-fcas` option. There will be a CAS there's
  another option that requires one.
- By default, it's an in-memory CAS.
- If `-fcas-builtin-path` is specified (and not empty), then it'll be an
  on-disk CAS.
- The other options were moved to PreprocessorOptions, FrontendOptions,
  and FileSystemOptions, as appropriate.

The other changes include:

- Renamed `-fcas-fs-result-cache` to `-fexperimental-cache-compile`.
- Renamed `-Rcas-fs-result-cache*` to `-Rcache-compile*`.
- Renamed `-fcas-token-cache` to `-fexperimental-cache-lex-raw`.
- Fixed up some old cruft in DEMO.md
- Partly cleaned up layering of "who creates the CAS, and how do you
  reference it?". `createVFSForCompilerInvocation()`, which might need
  to create a CAS, makes this a bit awkward.
    - Removed `createCASForCompilerInvocation()`, which had a global
      cache.
    - Added `CASOptions::getOrCreateCAS()`, which caches within
      the CASOptions instance (one per compiler invocation).
    - Added `CompilerInstance::getOrCreateCAS()`, which looks inside
      the VFS before calling `CASOptions`. It should work?
    - Needs more follow-ups in clangToolingDependencyScanning to pass
      in the current CAS. This will be easier if we change from
      `shared_ptr<CASDB>` to IntrusiveRefCntPtr, since that makes it
      easy to create owned references when you have `CASDB&`.
- New alias for the default CAS path: `-`.

I'm not sure about the new alias for the default CAS path (`-`).

- This is what typically gets used for stdin or stdout. That's
  attractive, since we already expect it to be "special". Also maybe
  surprising?
- I like that it's short.
- I had thought we could use `-` to mean "in-memory CAS" in llvm-cas and
  llvm-cas-object-format. Maybe we should just change those to make
  `--cas` optional and use an in-memory CAS by default? We'd want to add
  good errors in that case (on failure: "note: did you forget --cas?").

dexonsmith committed:
Clang: Substantially rework CAS-related command-line options

This substantially reworks the command-line options in Clang related to
the CAS, primarly in order to strip down `CASOptions` to only contain
CAS configuration, and not indications of how to use it. It also
fine-tuned how to configure the CAS, which caused a long-tail of things
layering problems to be pulled in. I cut off a few of the threads with
FIXMEs for now.

The new CASOptions only has a single field, BuiltinPath, which is stored
in a sub-structure called CASConfiguration.

- There is no longer an `-fcas` option. There will be a CAS there's
  another option that requires one.
- By default, it's an in-memory CAS.
- If `-fcas-builtin-path` is specified (and not empty), then it'll be an
  on-disk CAS.
- The other options were moved to PreprocessorOptions, FrontendOptions,
  and FileSystemOptions, as appropriate.

The other changes include:

- Renamed `-fcas-fs-result-cache` to `-fexperimental-cache-compile`.
- Renamed `-Rcas-fs-result-cache*` to `-Rcache-compile*`.
- Renamed `-fcas-token-cache` to `-fexperimental-cache-lex-raw`.
- Fixed up some old cruft in DEMO.md
- Partly cleaned up layering of "who creates the CAS, and how do you
  reference it?". `createVFSForCompilerInvocation()`, which might need
  to create a CAS, makes this a bit awkward.
    - Removed `createCASForCompilerInvocation()`, which had a global
      cache.
    - Added `CASOptions::getOrCreateCAS()`, which caches within
      the CASOptions instance (one per compiler invocation).
    - Added `CompilerInstance::getOrCreateCAS()`, which looks inside
      the VFS before calling `CASOptions`. It should work?
    - Needs more follow-ups in clangToolingDependencyScanning to pass
      in the current CAS. This will be easier if we change from
      `shared_ptr<CASDB>` to IntrusiveRefCntPtr, since that makes it
      easy to create owned references when you have `CASDB&`.
- New alias for the default CAS path: `-`.

I'm not sure about the new alias for the default CAS path (`-`).

- This is what typically gets used for stdin or stdout. That's
  attractive, since we already expect it to be "special". Also maybe
  surprising?
- I like that it's short.
- I had thought we could use `-` to mean "in-memory CAS" in llvm-cas and
  llvm-cas-object-format. Maybe we should just change those to make
  `--cas` optional and use an in-memory CAS by default? We'd want to add
  good errors in that case (on failure: "note: did you forget --cas?").

dexonsmith commented:
@swift-ci test

cachemeifyoucan commented:
> I'm not sure about the new alias for the default CAS path (-).

I am not sure if this is a good idea or not since it overloads what `-` means. This is also a `cc1` option and you never type cc1 option yourself. 
I do like how the option works now better.

> Renamed -fcas-fs-result-cache to -fexperimental-cache-compile.
> Renamed -Rcas-fs-result-cache* to -Rcache-compile*.
> Renamed -fcas-token-cache to -fexperimental-cache-lex-raw.

`-fexperimental` prefix is good but I feel like it should be a noun afterwards, not an action. 

Generally looks good.

benlangmuir commented:
I agree with @cachemeifyoucan on the option spelling:  The -f options sound better as nouns, and I think using "-" for something other than stdin is confusing in a command-line option.

> By default, it's an in-memory CAS.

For cc1, or for a testing tool this may be fine, but I would expect the clang driver to default to a persistent CAS if I'm using any kind of inter-compilation caching options, or else it could require me to be explicit about the CAS I want.

benlangmuir commented on clang/include/clang/Driver/Options.td:
```diff
@@ -6167,56 +6174,43 @@ def dump_depscan_tree_EQ : Option<["-"], "dump-depscan-tree=", KIND_JOINED>,
-defm fcasid_output : BoolFOption<"casid-output",
-    CASOpts<"WriteOutputAsCASID">, DefaultFalse,
-    PosFlag<SetTrue, [], "When using -fcas-fs-result-cache, write a CASID for the output file.">,
+def fcas_token_cache : Flag<["-"], "fcas-token-cache">,
```
Why are you keeping this alias for the token cache, but not any of the other modified options?

dexonsmith replied:
Because this `-cc1` is used by the CMake options. None of the others is. Thought I'd do it in stages:
- Introduce new function leave behind alias.
- Wait for that to be around for people (us) to have time to switch to a toolchain that has the new option.
- Change CMake to use the new option and remove the alias.

benlangmuir commented on llvm/cmake/modules/HandleLLVMOptions.cmake:
```diff
@@ -1302,8 +1302,10 @@ if(LLVM_ENABLE_EXPERIMENTAL_DEPSCAN)
-    "Cache tokens using -Xclang -fcas-cache-tokens")
+    "Cache tokens using -Xclang -fexperimental-cache-lex-raw")
 if(LLVM_ENABLE_EXPERIMENTAL_CAS_TOKEN_CACHE)
+  # FIXME: Update this to check -fexperimental-cache-lex-raw before dropping
```
Why not change it immediately?

dexonsmith replied:
Changing this will make it so that if you're using an old toolchain (i.e., from anything that doesn't have this commit) `SUPPORTS_CAS_TOKEN_CACHE` returns false.

EDIT: Effectively, that would break the feature temporarily. I'd rather not lose it.

benlangmuir commented on clang/lib/CAS/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+set(LLVM_LINK_COMPONENTS
+  Support
```
Shouldn't this depend on LLVM CAS as well? I think this would fail to link dynamically.

dexonsmith replied:
Fixed in e9567d33bfadb11309d954266554aca4d05977e3.

dexonsmith commented:
> > By default, it's an in-memory CAS.
> 
> For cc1, or for a testing tool this may be fine, but I would expect the clang driver to default to a persistent CAS if I'm using any kind of inter-compilation caching options, or else it could require me to be explicit about the CAS I want.

Yeah, makes sense to me; at this point it's only a `-cc1` option; and `-fdepscan` always overrides with the default on-disk CAS right now.

dexonsmith committed:
Add missing LLVMCAS dependency to clang/lib/CAS/CMakeLists.txt

dexonsmith commented:
From @cachemeifyoucan:
> I am not sure if this is a good idea or not since it overloads what `-` means. This is also a `cc1` option and you never type cc1 option yourself. I do like how the option works now better.
> 
> `-fexperimental` prefix is good but I feel like it should be a noun afterwards, not an action.

From @benlangmuir:
> I agree with @cachemeifyoucan on the option spelling: The -f options sound better as nouns, and I think using "-" for something other than stdin is confusing in a command-line option.

Naming ideas in either case?

For `-`:
- I'm fine with avoiding overloading stdin/stdout. It didn't quite feel right, but it seemed like a big improvement.
- While `-fcas-builtin-path` is currently a `-cc1` option, it *should* be a driver option. I think it should happen in the same commit that `-fdepscan` stops clobbering it, or just after (that's the only reason it's a `-cc1` option now: that it doesn't work how you'd expect it to).
- `/^llvm::cas::builtin::default/llvm.cas.builtin.default` does not seem like the right way for a user to specify the default on-disk location. What name should we use there? It doesn't have to be supported in the `-cc1` option.
    - `auto`
    - `default`
    - `-default-`
- For `-cc1`, we could use:
    - `-fcas-builtin-path-default` or `-fcas-builtin-path=<explicit-path>`
- It'd be really nice to have something for `llvm-cas` and `llvm-cas-object-format`.
    - Thoughts?

For the `-f` options:
- I don't like using `cache` as a noun that comes at the end:
    - `-fexperimental-raw-tokens-cache`
    - `-fexperimental-compilation-cache` (is that a good noun for what's being cached?)
- I prefer a prefix, which makes it an action again:
    - `-fexperimental-cache-raw-tokens`
    - `-fexperimental-cache-compilation`
- ... but the thing being cached is now a noun. Is that better? Or other ideas?
- An earlier version of the commit had `-fexperimental-cache-cc1`, which sounds great to me (way better than "compilation"), but I realized that won't make sense to people that don't work on the compiler. Thoughts?

cachemeifyoucan commented:
I am wondering which direction you want to take for builtin CAS path because I have seen two possibilities:

* With the current implementation, it has to be a cc1 only option when `-fdepscan` is used with a daemon because the CAS daemon talks to determines the CAS path and only the daemon should write that path. I guess we can make it a driver option but issue unused argument when daemon is used?
* Or we can let the clang driver to decide the CAS daemon talks to. Maybe we can create the daemon prefix by hash of parent PID + CASOption so clang driver will spawn a correct daemon when CAS option is changed?

benlangmuir commented:
> While -fcas-builtin-path is currently a -cc1 option, it should be a driver option.

Does it need to support an _explicit_ default in the driver? I can't think of any path options in clang that have an explicit default like that, but I can think of a few that don't: `-fmodules-cache-path`, `-fuse-ld`, and `--ld-path` all have an implicit default value, but you can't say e.g. `--ld-path=default`. The closest you can do is `--ld-path=ld`, but that would not match the default on a platform with a different default linker name.

Slightly different question: do we need to say "builtin"? I'm not sure what it's adding here.

> It'd be really nice to have something for llvm-cas and llvm-cas-object-format.

Same question I guess, why does it need an explicit way to request the default?

> /^llvm::cas::builtin::default/llvm.cas.builtin.default does not seem like the right way for a user to specify the default on-disk location. 

I agree, but if you think it's important to have a way to spell the default path explicitly, then this seems like a fine way to do it other than the fact it's way too long.  `/^llvm::cas::default` seems fine to me, for instance.

>  I don't like using cache as a noun that comes at the end:
> -fexperimental-raw-tokens-cache

I like this one 😅 

Would "caching" instead of "cache" help?  If not I'm not sure what to suggest, because I prefer these names that end with "-cache" 😺 

> -fexperimental-compilation-cache (is that a good noun for what's being cached?)

Are you saying that "compilation" would be misleading?  I'm not sure what the issue with this one is.

cachemeifyoucan commented:
> From @cachemeifyoucan:
> 
> > I am not sure if this is a good idea or not since it overloads what `-` means. This is also a `cc1` option and you never type cc1 option yourself. I do like how the option works now better.
> > `-fexperimental` prefix is good but I feel like it should be a noun afterwards, not an action.
> 
> From @benlangmuir:
> 
> > I agree with @cachemeifyoucan on the option spelling: The -f options sound better as nouns, and I think using "-" for something other than stdin is confusing in a command-line option.
> 
> Naming ideas in either case?
> 
> For `-`:
> 
> * I'm fine with avoiding overloading stdin/stdout. It didn't quite feel right, but it seemed like a big improvement.
> * While `-fcas-builtin-path` is currently a `-cc1` option, it _should_ be a driver option. I think it should happen in the same commit that `-fdepscan` stops clobbering it, or just after (that's the only reason it's a `-cc1` option now: that it doesn't work how you'd expect it to).
> * `/^llvm::cas::builtin::default/llvm.cas.builtin.default` does not seem like the right way for a user to specify the default on-disk location. What name should we use there? It doesn't have to be supported in the `-cc1` option.
>   
>   * `auto`
>   * `default`
>   * `-default-`
> * For `-cc1`, we could use:
>   
>   * `-fcas-builtin-path-default` or `-fcas-builtin-path=<explicit-path>`
> * It'd be really nice to have something for `llvm-cas` and `llvm-cas-object-format`.
>   
>   * Thoughts?
> 

maybe `@default`, `@memory`? A little bit more typing. 

> For the `-f` options:
> 
> * I don't like using `cache` as a noun that comes at the end:
>   
>   * `-fexperimental-raw-tokens-cache`
>   * `-fexperimental-compilation-cache` (is that a good noun for what's being cached?)
> * I prefer a prefix, which makes it an action again:
>   
>   * `-fexperimental-cache-raw-tokens`
>   * `-fexperimental-cache-compilation`
> * ... but the thing being cached is now a noun. Is that better? Or other ideas?
> * An earlier version of the commit had `-fexperimental-cache-cc1`, which sounds great to me (way better than "compilation"), but I realized that won't make sense to people that don't work on the compiler. Thoughts?

`-fexperimental-cached-raw-tokens` and `-fexperimental-cached-build`? I don't have anything better. English is not what I am good at.

benlangmuir commented:
> maybe `@default`, `@memory`? A little bit more typing.

`-fcas-builtin-path @default` wouldn't work, because it would be treated as a response file.  It would work with a joined argument (`-fcas-builtin-path=@default`), but I don't like adding new joined-only options.

> -fexperimental-cached-raw-tokens and -fexperimental-cached-build? I don't have anything better. English is not what I am good at.

My preference for the grammatical form would be `raw-tokens-cache` > `cached-raw-tokens` > `cache-raw-tokens`.

dexonsmith commented:
> > While -fcas-builtin-path is currently a -cc1 option, it should be a driver option.
> 
> Does it need to support an _explicit_ default in the driver? I can't think of any path options in clang that have an explicit default like that, but I can think of a few that don't: `-fmodules-cache-path`, `-fuse-ld`, and `--ld-path` all have an implicit default value, but you can't say e.g. `--ld-path=default`. The closest you can do is `--ld-path=ld`, but that would not match the default on a platform with a different default linker name.

There's a difference between "default CAS" and "default location for the builtin on-disk CAS".

Ultimately, I expect the default behaviour for clang (when not doing `-fdepscan` or similar) will be to use an in-memory CAS / action-cache / etc., with nothing being shared between processes unless a driver option says "please share between processes".

I think there should be a way to say "I want to use the on-disk CAS in a default location".

> Slightly different question: do we need to say "builtin"? I'm not sure what it's adding here.

Historically, opposed to a path to a plugin. Maybe the name is not great though.

It was: `-fcas={builtin,none,plugin}` and this option was `-fcas-builtin-path=....`; then, "builtin" made lots of sense. IIRC, without a path, it assumed "in-memory".

Assume that we have plugins. What should `-fcas-path=` mean? And what would this option be called?

> > It'd be really nice to have something for llvm-cas and llvm-cas-object-format.
> 
> Same question I guess, why does it need an explicit way to request the default?

These tools error by-default if `--cas` is not explicitly specified. We could change it, but IMO, it's the right behaviour since they're meant for debugging... it's important for someone to specify them explicitly.

> 
> > /^llvm::cas::builtin::default/llvm.cas.builtin.default does not seem like the right way for a user to specify the default on-disk location.
> 
> I agree, but if you think it's important to have a way to spell the default path explicitly, then this seems like a fine way to do it other than the fact it's way too long. `/^llvm::cas::default` seems fine to me, for instance.

That could be awkward on windows, since it's not a valid windows path. It has a root name but no root directory.

Maybe it could be a URL: `llvmcas://default` .

> 
> > I don't like using cache as a noun that comes at the end:
> > -fexperimental-raw-tokens-cache
> 
> I like this one 😅

If you like them, maybe they're fine. I guess it lines up with `-fmodules-cache`, which I forgot about.

> > -fexperimental-compilation-cache (is that a good noun for what's being cached?)
> 
> Are you saying that "compilation" would be misleading? I'm not sure what the issue with this one is.

Compilation is not at all specific; everything the compiler does is compilation; it feels imprecise to me, as it would apply to any type of caching we added to the compiler; feels tantamount to `-fexperimental-cache`.

(Whereas `-cc1` doesn't have this ambiguity. That's the cached action.)

dexonsmith commented:
Talking more about compilation / actions / data:
> My preference for the grammatical form would be `raw-tokens-cache` > `cached-raw-tokens` > `cache-raw-tokens`.

Note that `raw-tokens` are specific data, whereas `compilation` is a (non-specific) action. This mismatch is what led me to `lex-raw`.

This also got me thinking more about how the names came about and what they really mean. I'm not in love with the names I have in this patch but I do want to push back on using `cache` in these options as a noun, as if there are multiple caches.

There's a single CAS where all results will be stored, and a single cache where the actions that produce them will (optionally) be cached. For this cache:
- The key is a hash of the "action + arguments". (similar to request + arguments in the request evaluator)
- The value is a CAS ID, corresponding to the storage for "some result".

I originally called it a result cache. Everyone else seems to call it an action cache, and I guess I've mostly adjusted to that.

Given that context, there's some cognitive dissonance for me when talking about the "raw-tokens-cache" as a noun. It makes sense for `-fmodules-cache(-path)` because there is a thing that is the modules cache, and it exists at some path on disk. But there isn't a raw-tokens cache. There's just a single action cache. It doesn't even store the tokens—the tokens are in the CAS—it just caches the lexing action that produces the tokens.

dexonsmith committed:
Merge remote-tracking branch 'origin/experimental/cas/main' into HEAD

dexonsmith committed:
CAS: Address review comments for command-line options

- Set stage for `-fexperimental-cache=raw-lex,compile-job,etc.` in the
  driver but for now just name `-cc1` flags in a way that corresponds to
  that.
    - Rename diagnostics to match.
- For configuring the CAS:
    - In-memory CAS: `-fno-cas-path` or `-fcas-path=`
    - On-disk CAS in user's system cache: `-fcas-path=auto`
    - On-disk CAS in a specific location: `-fcas-path=path/to/cas`
- Add FIXMEs explaining what we're waiting for in various places.
- Also apply new CAS configuration options to `--cas` option in
  `llvm-cas` and `llvm-cas-object-format`.

dexonsmith commented:
Major update in d30f03e3ffec08e908027d9e4eb677e47a65215b (see that commit message for details). Also not sure if lld needs an update; just now running tests locally to see. @akyrtzi , are there any lld tests that rely on the "default" CAS path working? If not, maybe we should add one.

@swift-ci test

dexonsmith committed:
Fix a dangling references to -fcas-builtin-path

dexonsmith commented:
> Major update in [d30f03e](https://github.com/apple/llvm-project/commit/d30f03e3ffec08e908027d9e4eb677e47a65215b) (see that commit message for details). Also not sure if lld needs an update; just now running tests locally to see. @akyrtzi , are there any lld tests that rely on the "default" CAS path working? If not, maybe we should add one.

There don't seem to be, since the lld tests pass locally, and I didn't update `lld` or `llvm-libtool-darwin` to have any logic for a default CAS path.

c966b3f9b319dba15b805d56a633ca82ca39885c just updates a comment.

@swift-ci test

dexonsmith commented:
@swift-ci test

dexonsmith commented:
@cachemeifyoucan , @akyrtzi , @benlangmuir , are any of you living on `lld` and/or `llvm-libtool-darwin`?
- If so, I should probably update them to have matching command-line options and correctly access the "default" on-disk CAS in this PR.
- Else, I'm not sure I have an opinion—@akyrtzi, you're kind of code owner for those changes.... let me know what you think we should do.

akyrtzi commented:
These tools just pass a path option directly to `llvm::cas::createOnDiskCAS()`, we could update them later, post this PR, to treat "auto" specially, as clang will be doing.

dexonsmith committed:
Revert changes to DEMO.md for lld examples

benlangmuir commented on DEMO.md:
```diff
@@ -308,7 +308,7 @@ directory.
   find "$OBJECTSDIR"/lib/Support -name "*.o" |
   sort >objects-to-ingest &&
-  time "$BUILDDIR"/llvm-cas-object-format --cas "$TMPDIR/casfs.default" \
+  time "$BUILDDIR"/llvm-cas-object-format --cas "-" \
```
"auto"?

dexonsmith replied:
fixed in db299193ff636da2af57bca5751baa9f5260b2fa

benlangmuir commented on DEMO.md:
````diff
@@ -320,7 +320,7 @@ serialization of LinkGraph and works for both MachO and ELF. You can try use
 
 ```
-  time "$BUILDDIR"/llvm-cas-object-format --cas "$TMPDIR/casfs.default" \
+  time "$BUILDDIR"/llvm-cas-object-format --cas "-" \
````
"auto"?

dexonsmith replied:
fixed in db299193ff636da2af57bca5751baa9f5260b2fa

benlangmuir commented on DEMO.md:
````diff
@@ -336,15 +336,15 @@ Some other use flags to use:
 ```
-% llvm-cas -cas "$TMPDIR/casfs.default" --merge /path/to/dir > casid
-% ld64.lld -arch x86_64 -platform_version macos 12 12 --fcas-builtin-path "$TMPDIR/casfs.default" --fcas-fs @casid /path/to/dir/main.o -o a.out
+% llvm-cas -cas "-" --merge /path/to/dir > casid
````
"auto"?

dexonsmith replied:
Reverted this change due to lld piece in cf87e3c8ad973ca3adca33c21f28531c05c5a21e

benlangmuir commented on DEMO.md:
````diff
@@ -354,6 +354,6 @@ to indicate whether there was a cache 'hit' or 'miss'.
 ```
-% llvm-cas-object-format --cas "$TMPDIR/casfs.default" t1.o t2.o -silent > casid
-% ld64.lld -arch x86_64 -platform_version macos 12 12 --fcas-builtin-path "$TMPDIR/casfs.default" @casid -o a.out
+% llvm-cas-object-format --cas "-" t1.o t2.o -silent > casid
````
"auto"?

dexonsmith replied:
Reverted this change due to lld piece in cf87e3c8ad973ca3adca33c21f28531c05c5a21e

benlangmuir commented on clang/include/clang/Driver/Options.td:
````diff
@@ -6060,29 +6060,36 @@ def fsycl_is_host : Flag<["-"], "fsycl-is-host">,
+// For building against some pre-ingested, CAS-based filesystem in the builtin
+// on-disk CAS:
+// ```
+// % clang -fcas=on-disk -fcas-fs=llvmcas://1af12fa2afa1af2a1fa
````
`-fcas=on-disk` has reappeared in the comment, probably a mistake?

dexonsmith replied:
fixed in b349b5b897294705b65da1c68e904f95ae509b8d

benlangmuir commented on llvm/cmake/modules/HandleLLVMOptions.cmake:
```diff
@@ -1240,7 +1240,7 @@ set(LLVM_ENABLE_EXPERIMENTAL_DEPSCAN OFF CACHE BOOL
 set(LLVM_DEPSCAN_DAEMON "" CACHE STRING "Path to existing DepScan daemon to use")
 set(LLVM_CAS_BUILTIN_PATH "" CACHE STRING "Path to pass for -fcas-builtin-path")
-set(LLVM_CAS_BUILTIN_PATH_Default "/^llvm::cas::builtin::default/llvm.cas.builtin.default")
+set(LLVM_CAS_BUILTIN_PATH_Default "-")
```
"auto"?

dexonsmith replied:
Reverted in 7756976fa6d4cf23fbb7f124ba95003add6a106a since it gets passed to lld.

dexonsmith committed:
Revert change to cas::createOnDiskCAS()

Avoid changing `cas::createOnDiskCAS()` for now, replacing with FIXMEs,
to avoid breaking non-clang clients for now.

dexonsmith committed:
Remove the new `-fcas-path=` and `-fno-cas-path` aliases

Testing reveals they don't work, and we only want them
in the driver anyway. Just document the plan.

Found via clang/test/Frontend/cc1-return-codes.c, which caught that
```
% clang -cc1 -help
```
had started crashing on an assertion to do with `AliasArgs`.

dexonsmith committed:
Fix another couple of `--cas=auto`s

dexonsmith committed:
Fix Example command-line in Options.td comments

dexonsmith committed:
Revert to `/^llvm::cas::builtin::default/llvm.cas.builtin.default` for LLVM_CAS_BUILTIN_PATH_Default

This goes to lld and clang, and lld does not have `auto` yet,
so cannot move to `auto`.

dexonsmith commented:
FYI, there was a test failure, which 680e8b6e37efe4f10fedfb1f32b3d3145f4e7ff8 fixed. It removed the `-fcas-path=` and `-fno-cas-path` aliases since they weren't working and were breaking `clang -cc1 -help`. We only need/want those in the driver anyway.

dexonsmith commented:
@swift-ci test
